### ISSUE TITLE
Feature requests: calculate from & to, and calculate reference integers

### ISSUE THREAD

**Issue by jeff-h** [TechWilk/bible-verse-parser#24]
Nice work on this! Thanks for sharing it.

I'm wondering if you'd consider a couple of feature requests to provide a little more info about each passage:

* parse out a "from" and a "to" reference for each passage eg `John 3:16-18` would provide `John 3:16` and `John 3:18`
* provide the "from" and "to" in integer format  as per below eg the above would provide something like `43003016` and `43003018`

It seems somewhat common to convert a single verse reference into an integer like so:

[(1000000 * Book.toInt book) + (1000 * chapter) + verse](https://github.com/monty5811/elm-bible/blob/2084c58f4c0849d06020c8bd4d6a8a07271ed63b/src/Bible.elm#L31)

This format allows for some really simple & fast SQL searches and such eg "return all the passages which contain the verse John 3:17" or "return all the passages which are in Genesis 17"

Examples of this idea:
* https://github.com/scrollmapper/bible_databases#verse-id-system
* https://package.elm-lang.org/packages/monty5811/elm-bible/latest/Bible#Encoding ("Encoding" section)

These requests would require a source of book metadata eg [this](https://github.com/scrollmapper/bible_databases/blob/963ff9cf716f13fa20dd0982204b86a878a29c54/json/key_english.json).

This would also facilitate items 2, 3 and 4 on your roadmap :)

**Comment by TechWilk** [comment]
Thanks!

Yes those suggestions make a lot of sense. As you can see, the package started off just to split up multiple passages, but didn't need to _understand_ the structure of the passage at the start.

I've been recently starting to think again about how to properly handle verses spanning across chapters (recently requested again in #18) and was starting to come to similar conclusions that I'll need to move to a from/to pattern. Your thoughts have also helped confirm that! With this in place I should be able to hit the rest of the roadmap in one go.

Something along the lines of this, once parsed:

```php
$passage
  ->from() // $reference
  ->to() // $reference
  ->all() // array of references, including from, to & all between
  ->toString() // string "John 3:16-18"

$reference
  ->book() // "John"
  ->abbreviatedBook() // "Jn"
  ->chapter() // 3
  ->verse() // 16
  ->fragment() // null (a/b/c if used)
  ->integerNotation() // 43003016
  ->toString() // "John 3:16"
```

The integer format is pretty cool! I'd not come across that before. It's definitely something this package should support.
One thing I see that it does omit is sub-verse fragments, such as `Genesis 14:20a`. I don't see them used often, but do occasionally.

So, I've got a branch in process tackling roadmap items 2/3 and introducing some book metadata which provides a starting point for this, so it makes sense to get that merged first before turning . I'll target the following versions:
- `v1.2.0` book name validation to the existing single-chapter passages, introducing book metadata
- `v2.0.0` from/to passages which can span across chapters or even books. A breaking change to introduce the new concepts

It will probably take the next few weeks on-and-off for me to find the time for these. Are you working on a project with particular time sensitivity?

**Comment by TechWilk** [comment]
Apologies for the delay.

Yes those are perfectly valid suggestions.

1. It's true that the `->all()` array could become pretty big. I think it's useful though since the integer format doesn't allow for computational analysis - it doesn't allow 
2. Agreed, it seems reasonable that a chapter is also represented as a from/to of the first and last verse. It'll definitely be a bit more work, but I think it'll be worth it.
3. Very prudent. Yes, the branch isn't anywhere near production code yet, it's much more of a prototype. I was planning to rearrange how the data is stored to a format near-identical to the one you give as an example. It was flat only to make it easy to combine multiple abbreviation datasets with crude bash scripts!

Thanks very much for doing the work to improve the data structure! I'll try and find time this week to move the parser along a bit

**Comment by TechWilk** [comment]
I've released `v1.2.0` which includes support for abbreviations & checks book names are valid from the new data structure. 

Hopefully I'll have some more time this week to look into passages which span chapter across boundaries (ping #18)

**Comment by jeff-h** [comment]
This is fantastic! I've been distracted by other priorities unfortunately but I am definitely looking forward to circling back to this. I will let you know how it goes.
